pkg/controller/jobs/rayjob/rayjob_controller.go

```diff
@@ -0,0 +1,278 @@
+			workerPodSpec.NodeSelector = cloneSelectors(nodeSelectors[index+1].NodeSelector)
+		}
+	}
+}
+
+func (j *RayJob) Finished() (metav1.Condition, bool) {
+	condition := metav1.Condition{
+		Type:    kueue.WorkloadFinished,
+		Status:  metav1.ConditionTrue,
+		Reason:  string(j.Status.JobStatus),
+		Message: j.Status.Message,
+	}
+	return condition, rayjobapi.IsJobTerminal(j.Status.JobStatus)
+}
+
+func (j *RayJob) EquivalentToWorkload(wl kueue.Workload) bool {
+	if len(wl.Spec.PodSets) != (len(j.Spec.RayClusterSpec.WorkerGroupSpecs) + 1) {
+		return false
+	}
+
+	podSets := wl.Spec.PodSets
+	headPodSpec := &j.Spec.RayClusterSpec.HeadGroupSpec.Template.Spec
+	if podSets[0].Count != 1 {
+		return false
+	}
+	if !equality.Semantic.DeepEqual(headPodSpec.InitContainers, podSets[0].Template.Spec.InitContainers) {
+		return false
+	}
+	if !equality.Semantic.DeepEqual(headPodSpec.Containers, podSets[0].Template.Spec.Containers) {
+		return false
+	}
+
+	// workers
+	for index := range j.Spec.RayClusterSpec.WorkerGroupSpecs {
+		// MinReplicas and MaxReplicas should be taken into account for partial admission.
+		if podSets[index+1].Count != *j.Spec.RayClusterSpec.WorkerGroupSpecs[index].Replicas {
+			return false
+		}
+
+		workerPodSpec := &j.Spec.RayClusterSpec.WorkerGroupSpecs[index].Template.Spec
+		if !equality.Semantic.DeepEqual(workerPodSpec.InitContainers, podSets[index+1].Template.Spec.InitContainers) {
+			return false
+		}
+		if !equality.Semantic.DeepEqual(workerPodSpec.Containers, podSets[index+1].Template.Spec.Containers) {
+			return false
+		}
+	}
+	return true
+}
+
+func (j *RayJob) PriorityClass() string {
+	if j.Spec.RayClusterSpec != nil {
+		rcs := j.Spec.RayClusterSpec
+		if len(rcs.HeadGroupSpec.Template.Spec.PriorityClassName) != 0 {
+			return rcs.HeadGroupSpec.Template.Spec.PriorityClassName
+		}
+		for wi := range rcs.WorkerGroupSpecs {
+			w := &rcs.WorkerGroupSpecs[wi]
+			if len(w.Template.Spec.PriorityClassName) != 0 {
+				return w.Template.Spec.PriorityClassName
+			}
+		}
+	}
+	return ""
+}
+
+func (j *RayJob) PodsReady() bool {
+	return j.Status.RayClusterStatus.State == rayjobapi.Ready
+}
+
+func (j *RayJob) ReclaimablePods() []kueue.ReclaimablePod {
+	return nil
+}
+
+// SetupWithManager sets up the controller with the Manager. It indexes workloads
+// based on the owning jobs.
+func (r *RayJobReconciler) SetupWithManager(mgr ctrl.Manager) error {
+	return ctrl.NewControllerManagedBy(mgr).
+		For(&rayjobapi.RayJob{}).
+		Owns(&kueue.Workload{}).
+		Complete(r)
+}
+
+func SetupIndexes(ctx context.Context, indexer client.FieldIndexer) error {
+	return jobframework.SetupWorkloadOwnerIndex(ctx, indexer, gvk)
+}
+
+//+kubebuilder:rbac:groups="",resources=events,verbs=create;watch;update
+//+kubebuilder:rbac:groups=ray.io,resources=rayjobs,verbs=get;list;watch;update;patch
+//+kubebuilder:rbac:groups=ray.io,resources=rayjobs/status,verbs=get
```

Comment:
```suggestion
//+kubebuilder:rbac:groups=ray.io,resources=rayjobs/status,verbs=get;update
```

I found we need to add permission to update statuses to framework job controllers.

ref: #797

Reply:
thanks, done